### PR TITLE
introducing connection timeout and query read timeout

### PR DESCRIPTION
**Pull request by jaroslawk**
Alaisi,

We are using your driver in our app under AWS cloud and we are missing some failover scenarios support:

- connection timeout - max wait time for connecting to the DB. Without it we wait forever for the response from DB and there is no way for us to signal error to RX side.
- read timeout - this is similar situation. On the network issues one need to be prepare to have no feedback packet when attempting to use existing connection.

I have provided you the tests for those scenarios. 

Please let me know what do you think about those changes?

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -93,6 +93,12 @@
             <scope>test</scope>
         </dependency>
         <dependency>
+            <groupId>ch.qos.logback</groupId>
```
Comment (alaisi) [inline pom.xml]: Why the logback dependency?

File: src/main/java/com/github/pgasync/impl/PgConnectionPool.java [hunk 2]
```diff
@@ -178,8 +178,7 @@ public void close() throws Exception {
                     lock.unlock();
                 }
             }
-        }).flatMap(conn -> validator.call(conn).doOnError(err -> release(conn)))
-                .retry(poolSize + 1);
+        }).flatMap(conn -> validator.call(conn).doOnError(err -> release(conn)));
```
Comment (alaisi) [inline src/main/java/com/github/pgasync/impl/PgConnectionPool.java]: Is removing the retry call intentional? Why?

File: src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java [hunk 3]
```diff
@@ -115,10 +119,23 @@ public NettyPgProtocolStream(EventLoopGroup group, SocketAddress address, boolea
...
+            initReadTimeout(subscriber);
 
         }).lift(throwErrorResponsesOnComplete());
     }
 
+    private void initReadTimeout(Subscriber subscriber) {
+        if (readTimeout != 0) {
+            Runnable onTimeout = () -> {
+                ReadTimeoutException timeoutException = ReadTimeoutException.INSTANCE;
+                subscriber.onError(timeoutException);
```
Comment (alaisi) [inline src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java]: Won't this lead to duplicate onError calls? I would assume fireExceptionCaught would propagate the exception to all subscribers.

File: src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java [hunk 4]
```diff
@@ -250,14 +270,27 @@ public void userEventTriggered(ChannelHandlerContext context, Object evt) throws
...
+                channel.pipeline().addLast(newExceptionHandler());
+            }
+        };
+    }
+
+    private ChannelInboundHandlerAdapter newExceptionHandler() {
+        return new ChannelInboundHandlerAdapter() {
+            @Override
+            public void exceptionCaught(ChannelHandlerContext context, Throwable cause) throws Exception {
+                Collection<Subscriber<? super Message>> unsubscribe = subscribers;
```
Comment (alaisi) [inline src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java]: This looks fishy. As removeAll is called with itself, unsubscribe is always empty and no onError calls are performed. The original version looks wrong as well, this should probably be something like:

```java
if(!subscribers.isEmpty()) {
    Collection<Subscriber<? super Message>> unsubscribed = new ArrayList<>(subscribers);
    subscribers.clear(); // clear subs before onError to prevent recursive invocation
    unsubscribed.forEach(subscriber -> {
        try {
            subscriber.onError(cause)
        } catch(Exception ex) { /* ignored */ }
    });
}
```

File: src/test/resources/log4j2-test.xml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (alaisi) [inline src/test/resources/log4j2-test.xml]: Is log4j2 is use?

File: src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java [hunk 6]
```diff
@@ -250,14 +270,27 @@ public void userEventTriggered(ChannelHandlerContext context, Object evt) throws
...
-                if(useSsl) {
+                if (useSsl) {
                     channel.pipeline().addLast(newSslInitiator());
                 }
                 channel.pipeline().addLast(new LengthFieldBasedFrameDecoder(Integer.MAX_VALUE, 1, 4, -4, 0, true));
                 channel.pipeline().addLast(new ByteBufMessageDecoder());
                 channel.pipeline().addLast(new ByteBufMessageEncoder());
                 channel.pipeline().addLast(newProtocolHandler());
                 channel.pipeline().addLast(onActive);
+                channel.pipeline().addLast(newExceptionHandler());
```
Comment (alaisi) [inline src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java]: Why is this moved to a new handler?

File: pom.xml (hunk 1 above)
Comment (jaroslawk) [reply pom.xml]: this was just to see the netty detailed logs doing testing. it is not really needed.

File: src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java (hunk 6 above)
Comment (jaroslawk) [reply src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java]: I wanted to have exception caught regardless of anything and passed to subscriber, also separation of concerns

File: src/main/java/com/github/pgasync/impl/PgConnectionPool.java (hunk 2 above)
Comment (jaroslawk) [reply src/main/java/com/github/pgasync/impl/PgConnectionPool.java]: it was. retry logic looks like a separate concern to me. for a heavy analytic query which timeout you do not want to retry it. 

for short simple query for sure.

in summary this needs to be revisited.

File: src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java (hunk 3 above)
Comment (jaroslawk) [reply src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java]: initially it was wired like that. but for some reason I could not see this exception bubble up on exception handler.

File: src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java (hunk 4 above)
Comment (jaroslawk) [reply src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java]: better you can extract try{ } catch to separate method and also

```if(!subscribers.isEmpty())```
is not needed. forEach on empty list is harmless.

File: src/test/resources/log4j2-test.xml (hunk 5 above)
Comment (jaroslawk) [reply src/test/resources/log4j2-test.xml]: dirty way to get logging working.

**Comment by jaroslawk** [comment]
@alaisi  I am removing not important changes and also I see timeout was not working correctly.
This time it covers also initial message - this covers our usage.

Let me know what you think.

I will add utest to prove it works as expected.

File: src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java [hunk 7]
```diff
@@ -115,10 +123,25 @@ public NettyPgProtocolStream(EventLoopGroup group, SocketAddress address, boolea
...
         }).lift(throwErrorResponsesOnComplete());
     }
 
+    private ScheduledFuture<?> initReadTimeout(Subscriber subscriber) {
+        if (readTimeout != 0) {
+            Runnable onTimeout = () -> {
+                ReadTimeoutException timeoutException = ReadTimeoutException.INSTANCE;
+                subscriber.onError(timeoutException);
+                ctx.fireExceptionCaught(timeoutException);
+                ctx.close();
```
Comment (jaceksokol) [inline src/main/java/com/github/pgasync/impl/netty/NettyPgProtocolStream.java]: Closing this channel requires connection removal from the poll. Validator needs to be updated (it is in one of other PRs).

**Comment by jaroslawk** [comment]
Jacek - good point. I have updated the code to take only connections which are alive under consideration.
